core/server/src/main/java/alluxio/master/journal/AsyncJournalWriter.java

```diff
@@ -0,0 +1,141 @@
+  private final long mFlushBatchTime;
+
+  /**
+   * Use a {@link ReentrantLock} to guard the journal writing. Using the fairness policy seems to
+   * result in better throughput.
+   */
+  private final ReentrantLock mFlushLock = new ReentrantLock(true);
+
+  /**
+   * Creates a {@link AsyncJournalWriter}.
+   *
+   * @param journalWriter the {@link JournalWriter} to use for writing
+   */
+  public AsyncJournalWriter(JournalWriter journalWriter) {
+    mJournalWriter = Preconditions.checkNotNull(journalWriter);
+    mQueue = new ConcurrentLinkedQueue<>();
+    mCounter = new AtomicLong(0);
+    mFlushCounter = new AtomicLong(0);
+    mFlushBatchTime = MasterContext.getConf().getLong(Constants.MASTER_JOURNAL_FLUSH_BATCH_TIME_MS);
+  }
+
+  /**
+   * Appends a {@link JournalEntry} for writing to the journal.
+   *
+   * @param entry the {@link JournalEntry} to append
+   * @return a counter for the entry, for flushing
+   */
+  public long appendEntry(JournalEntry entry) {
+    // TODO(gpang): handle bounding the queue if it becomes too large.
+
+    /**
+     * Protocol for appending entries
+     *
+     * This protocol is lock free, to reduce the overhead in critical sections. It uses
+     * {@link AtomicLong} and {@link ConcurrentLinkedQueue} which are both lock-free.
+     *
+     * The invariant that must be satisfied is that the 'counter' that is returned must be
+     * greater than or equal to the actual counter of the entry in the queue.
+     *
+     * In order to guarantee the invariant, the {@link #mCounter} is incremented before adding the
+     * entry to the {@link #mQueue}. AFTER the counter is incremented, whenever the counter is
+     * read, it is guaranteed to be greater than or equal to the counter for the queue entries.
+     *
+     * Therefore, the {@link #mCounter} must be read AFTER the entry is added to the queue. The
+     * resulting read of the counter AFTER the entry is added is guaranteed to be greater than or
+     * equal to the counter for the entries in the queue.
+     */
+    mCounter.incrementAndGet();
+    mQueue.offer(entry);
+    return mCounter.get();
+  }
+
+  /**
+   * Flushes and waits until the specified counter is flushed to the journal. If the specified
+   * counter is already flushed, this is essentially a no-op.
+   *
+   * @param counter the counter to flush
+   * @throws IOException if an error occurs in flushing the journal
+   */
+  public void flush(final long counter) throws IOException {
+    if (counter <= mFlushCounter.get()) {
+      return;
+    }
+    // Using reentrant lock, since it seems to result in higher throughput than using 'synchronized'
+    mFlushLock.lock();
+    try {
+      long startTime = System.currentTimeMillis();
+      long flushCounter = mFlushCounter.get();
+      if (counter <= flushCounter) {
+        // The specified counter is already flushed, so just return.
+        return;
+      }
+      while (counter > flushCounter) {
```

Comment:
I think it is slightly different. With the inner loop, it will finish writing all entries that exist in the queue, essentially batch writing a lot. With only one while loop, it will stop exactly at the entry requested. The batching was originally intentional.